cpp/include/raft/neighbors/detail/cagra/cagra_search.cuh

```diff
@@ -173,6 +147,145 @@ void search_main(raft::resources const& res,
             topk,
             set_offset(sample_filter, qid));
   }
+}
+
+template <class T,
+          class DatasetT,
+          class DatasetIdxT,
+          class InternalIdxT,
+          class DistanceT,
+          class CagraSampleFilterT>
+void lauch_vpq_search_main_core(
+  raft::resources const& res,
+  const vpq_dataset<DatasetT, DatasetIdxT>* vpq_dset,
+  search_params params,
+  raft::device_matrix_view<const InternalIdxT, int64_t, row_major> graph,
+  raft::device_matrix_view<const T, int64_t, row_major> queries,
+  raft::device_matrix_view<InternalIdxT, int64_t, row_major> neighbors,
+  raft::device_matrix_view<DistanceT, int64_t, row_major> distances,
+  CagraSampleFilterT sample_filter)
+{
+  const float vq_scale = 1.0f;
+  const float pq_scale = 1.0f;
+
+  if (vpq_dset->pq_bits() == 8) {
+    if (vpq_dset->pq_len() == 2) {
+      using dataset_desc_t = cagra_q_dataset_descriptor_t<T,
+                                                          DatasetT,
+                                                          8 /*PQ bit*/,
+                                                          2 /* Subspace dimension*/,
+                                                          0,
+                                                          DistanceT,
+                                                          InternalIdxT,
+                                                          0>;
```

Comment:
Yes, right. They are replaced with actual values just before the search kernel is launched. This is because we can't determine `TEAM_SIZE` and `BLOCK_DIM` when making the instance of dataset_desc. But I'm not sure this code design is good.
https://github.com/enp1s0/raft/blob/38ab2bd787699992a23296532b770859b1bf7512/cpp/include/raft/neighbors/detail/cagra/search_single_cta_kernel-inl.cuh#L913-L914

Reply:
An alternative design is to remove the `TEAM_SIZE' and `BLOCK_DIM' template parameters from the dataset_desc and add them to the `compute_similarity' function instead.